fixes for compilation under MS Visual Studio 2015

jtrmal opened the pull request:
The compilation won't finish yet -- there is still a couple of the errors of the same kinds as the changeset is addressing. I do this just to solicit feedback from @danpovey (I'm gonna let him decide if he's willing to merge this) and perhaps @kkm000 

@ceefour, in the interim, you can apply the patch and see if it addresses some of your issues? 

y.

jtrmal commented:
BTW, it seems to me I see certain pattern emerging.
I was able to fix some issues by just doing explicit instantiation of
```
+// Explicit template instantiation to avoid name resolution conflicts
 +// with VS2015 -- preventing errors such as "ambiguous type uint64"
 +template fst::OLabelCompare<kaldi::LatticeArc>;
 +template fst::ILabelCompare<kaldi::LatticeArc>;
```

The is another quite common issue and its tied to ``ConvertLattice``
That is however more problematic, because during the expansion a "new" class with the template signature ``fst::ExpandedFst<fst::ArcTpl<fst::LatticeWeightTpl<double>>>`` gets generated and it fails because of the built-in FST registration procedure, which is mostly defunct in openfstwin and the linking fails:
```
1>lattice-lmrescore-const-arpa.obj : error LNK2001: unresolved external symbol "__declspec(dllimport) private: static int fst::GenericRegister<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,struct fst::FstRegisterEntry<class fst::ArcTpl<class fst::LatticeWeightTpl<double> > >,class fst::FstRegister<class fst::ArcTpl<class fst::LatticeWeightTpl<double> > > >::register_init_" (__imp_?register_init_@?$GenericRegister@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@U?$FstRegisterEntry@V?$ArcTpl@V?$LatticeWeightTpl@N@fst@@@fst@@@fst@@V?$FstRegister@V?$ArcTpl@V?$LatticeWeightTpl@N@fst@@@fst@@@4@@fst@@0HA)
1>lattice-lmrescore-const-arpa.obj : error LNK2001: unresolved external symbol "__declspec(dllimport) private: static class fst::Mutex * fst::GenericRegister<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,struct fst::FstRegisterEntry<class fst::ArcTpl<class fst::LatticeWeightTpl<double> > >,class fst::FstRegister<class fst::ArcTpl<class fst::LatticeWeightTpl<double> > > >::register_lock_" (__imp_?register_lock_@?$GenericRegister@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@U?$FstRegisterEntry@V?$ArcTpl@V?$LatticeWeightTpl@N@fst@@@fst@@@fst@@V?$FstRegister@V?$ArcTpl@V?$LatticeWeightTpl@N@fst@@@fst@@@4@@fst@@0PEAVMutex@2@EA)
1>lattice-lmrescore-const-arpa.obj : error LNK2001: unresolved external symbol "__declspec(dllimport) private: static class fst::FstRegister<class fst::ArcTpl<class fst::LatticeWeightTpl<double> > > * fst::GenericRegister<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,struct fst::FstRegisterEntry<class fst::ArcTpl<class fst::LatticeWeightTpl<double> > >,class fst::FstRegister<class fst::ArcTpl<class fst::LatticeWeightTpl<double> > > >::register_" (__imp_?register_@?$GenericRegister@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@U?$FstRegisterEntry@V?$ArcTpl@V?$LatticeWeightTpl@N@fst@@@fst@@@fst@@V?$FstRegister@V?$ArcTpl@V?$LatticeWeightTpl@N@fst@@@fst@@@4@@fst@@0PEAV?$FstRegister@V?$ArcTpl@V?$LatticeWeightTpl@N@fst@@@fst@@@2@EA)
```

danpovey commented:
I don't see why that registration would cause a linking a problem- that's
only there to enable reading to work.

It seems to me that the changes you have had to make are a bit excessive
and onerous-- forcing us to type 'kaldi::' everywhere seems a bit much.

I wonder if you could change the definitions of int32, int64, etc., to not
refer to the POSIX types but instead to refer via typedef to the
corresponding FST versions and #include the relevant header file from
OpenFst, might be easier.  I had avoided doing that because I wanted the
matrix library to be usable without the dependency on OpenFst, but it would
be OK if it was inside an #ifdef for _MSC_VER.


Dan


On Fri, Mar 18, 2016 at 12:14 PM, jtrmal <notifications@github.com> wrote:

> BTW, it seems to me I see certain pattern emerging.
> I was able to fix some issues by just doing explicit instantiation of
>
> +// Explicit template instantiation to avoid name resolution conflicts
>  +// with VS2015 -- preventing errors such as "ambiguous type uint64"
>  +template fst::OLabelCompare<kaldi::LatticeArc>;
>  +template fst::ILabelCompare<kaldi::LatticeArc>;
>
> The is another quite common issue and its tied to ConvertLattice
> That is however more problematic, because during the expansion a "new"
> class with the template signature
> fst::ExpandedFst<fst::ArcTpl<fst::LatticeWeightTpl<double>>> gets
> generated and it fails because of the built-in FST registration procedure,
> which is mostly defunct in openfstwin and the linking fails:
>
> 1>lattice-lmrescore-const-arpa.obj : error LNK2001: unresolved external symbol "__declspec(dllimport) private: static int fst::GenericRegister<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,struct fst::FstRegisterEntry<class fst::ArcTpl<class fst::LatticeWeightTpl<double> > >,class fst::FstRegister<class fst::ArcTpl<class fst::LatticeWeightTpl<double> > > >::register_init_" (__imp_?register_init_@?$GenericRegister@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@U?$FstRegisterEntry@V?$ArcTpl@V?$LatticeWeightTpl@N@fst@@@fst@@@fst@@V?$FstRegister@V?$ArcTpl@V?$LatticeWeightTpl@N@fst@@@fst@@@4@@fst@@0HA)
> 1>lattice-lmrescore-const-arpa.obj : error LNK2001: unresolved external symbol "__declspec(dllimport) private: static class fst::Mutex * fst::GenericRegister<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,struct fst::FstRegisterEntry<class fst::ArcTpl<class fst::LatticeWeightTpl<double> > >,class fst::FstRegister<class fst::ArcTpl<class fst::LatticeWeightTpl<double> > > >::register_lock_" (__imp_?register_lock_@?$GenericRegister@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@U?$FstRegisterEntry@V?$ArcTpl@V?$LatticeWeightTpl@N@fst@@@fst@@@fst@@V?$FstRegister@V?$ArcTpl@V?$LatticeWeightTpl@N@fst@@@fst@@@4@@fst@@0PEAVMutex@2@EA)
> 1>lattice-lmrescore-const-arpa.obj : error LNK2001: unresolved external symbol "__declspec(dllimport) private: static class fst::FstRegister<class fst::ArcTpl<class fst::LatticeWeightTpl<double> > > * fst::GenericRegister<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,struct fst::FstRegisterEntry<class fst::ArcTpl<class fst::LatticeWeightTpl<double> > >,class fst::FstRegister<class fst::ArcTpl<class fst::LatticeWeightTpl<double> > > >::register_" (__imp_?register_@?$GenericRegister@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@U?$FstRegisterEntry@V?$ArcTpl@V?$LatticeWeightTpl@N@fst@@@fst@@@fst@@V?$FstRegister@V?$ArcTpl@V?$LatticeWeightTpl@N@fst@@@fst@@@4@@fst@@0PEAV?$FstRegister@V?$ArcTpl@V?$LatticeWeightTpl@N@fst@@@fst@@@2@EA)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/600#issuecomment-198430907>
>

jtrmal commented:
I unified the kaldi [u]int* types with the fst types. Now everything seems to compile fine (running complete compilation overnight). Will tweak some more things, so please do not commit yet, even if the travis checks finishes OK (which it should).
y.

kkm000 commented on src/latbin/lattice-add-trans-probs.cc:
```diff
@@ -65,8 +65,6 @@ int main(int argc, char *argv[]) {
         lats_rspecifier = po.GetArg(2),
         lats_wspecifier = po.GetArg(3);
-    
-    int32 n_done = 0;
```
This is unintentional, is not it? The variable is used below.

jtrmal replied:
yes, carelessness on my side. Thanks

kkm000 commented on src/base/kaldi-types.h:
```diff
@@ -38,14 +38,26 @@ typedef float   BaseFloat;
-typedef int16_t         int16;
-typedef int32_t         int32;
-typedef int64_t         int64;
+#if _MSC_VER >= 1900
```
Maybe this should not even be conditional on MSC? The types would be correct either way, if my understanding is correct.

danpovey replied:
Yes, we discussed making it like this, Yenda will change it.


On Sat, Mar 19, 2016 at 10:45 PM, Kirill Katsnelson <
notifications@github.com> wrote:

> In src/base/kaldi-types.h
> <https://github.com/kaldi-asr/kaldi/pull/600#discussion_r56758524>:
>
> >
> >  namespace kaldi {
> > -typedef uint16_t        uint16;
> > -typedef uint32_t        uint32;
> > -typedef uint64_t        uint64;
> > -typedef int16_t         int16;
> > -typedef int32_t         int32;
> > -typedef int64_t         int64;
> > +#if _MSC_VER >= 1900
>
> Maybe this should not even be conditional on MSC? The types would be
> correct either way, if my understanding is correct.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/600/files/c2e0c05d693409ab30dfad03c56003fdba424e47#r56758524>
>

kkm000 replied:
By the way, [this part of OpenFSTWin patch](https://github.com/kaldi-asr/kaldi/blob/master/tools/extras/openfstwin-1.3.4.patch#L9) won't be needed, I think, with this change. I do not know if there is anything else Windows- or MSVC-specific in this patch. @jtrmal?

ceefour commented:
`kaldi-nnet3` compiled for me using this: :)

```
1>  Generating Code...
1>  Compiling...
1>  nnet-combine.cc
1>  nnet-chain-example.cc
1>  nnet-chain-diagnostics.cc
1>  nnet-chain-combine.cc
1>  nnet-analyze.cc
1>  nnet-am-decodable-simple.cc
1>  natural-gradient-online.cc
1>  am-nnet-simple.cc
1>  Generating Code...
1>  kaldi-nnet3.vcxproj -> D:\kaldi-trunk\kaldiwin_vs2015_MKL\x64\Debug\kaldi-nnet3.lib
========== Build: 1 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
```

kkm000 commented:
@ceefour, thanks for the feedback!

ceefour commented:
I've built the whole solution, seems pretty good to me. Thanks @jtrmal @kkm000 :+1: 

```
...
680>  weight-silence-post.cc
680>  weight-silence-post.vcxproj -> D:\kaldi-trunk\kaldiwin_vs2015_MKL\x64\Debug\weight-silence-post.exe
681>------ Build started: Project: word-align-lattice-lexicon-test, Configuration: Debug x64 ------
681>  word-align-lattice-lexicon-test.cc
681>  word-align-lattice-lexicon-test.vcxproj -> D:\kaldi-trunk\kaldiwin_vs2015_MKL\x64\Debug\word-align-lattice-lexicon-test.exe
========== Build: 670 succeeded, 11 failed, 0 up-to-date, 0 skipped ==========
```

I haven't checked the 11 failed, but it seems they're about `portaudio.h`.

https://gist.github.com/ceefour/013ecffec30bd07058c7

jtrmal commented:
I addressed the comments.  Kudos to @kkm000 for noticing the extra-definition of types in the patch.
Will try to apply the patchset on linux to see if it runs cleanly.
y.

kkm000 commented:
LGTM.

danpovey commented:
Looks OK to me.  We can now remove all the redundant typedefs of int32 and int64, but OK to do it in another patch. 
Dan

jtrmal commented:
This is ready to merge -- I squashed the commits and checked it compiles under linux

kkm000 commented:
LGTM. No, it looks *amazing* to me!

Can't wait for this as I am experimenting with building kaldi libraries with the original patched openfst (not openfstwin).

danpovey commented:
yenda, merge this yourself if it's ready- not sure if the issues you found
with double-precision compilation are related.
dan


On Tue, Mar 22, 2016 at 10:37 PM, Kirill Katsnelson <
notifications@github.com> wrote:

> LGTM. No, it looks *amazing* to me!
>
> Can't wait for this as I am experimenting with building kaldi libraries
> with the original patched openfst (not openfstwin).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/600#issuecomment-200138025>
>

jtrmal commented:
OK, merging.

jtrmal commented:
No, this is resolved by #600

kkm000 commented:
@jtrmal: you meant to comment on #597?

jtrmal commented:
yeah... wanted to delete it but you were faster :)
y.

On Tue, Mar 22, 2016 at 10:54 PM, Kirill Katsnelson <
notifications@github.com> wrote:

> @jtrmal <https://github.com/jtrmal>: you meant to comment on #597
> <https://github.com/kaldi-asr/kaldi/pull/597>?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/600#issuecomment-200143761>
>
